src/test/java/com/puppycrawl/tools/checkstyle/bdd/ParsedInput.java

```diff
@@ -0,0 +1,103 @@
+////////////////////////////////////////////////////////////////////////////////
+// checkstyle: Checks Java source code for adherence to a set of rules.
+// Copyright (C) 2001-2021 the original author or authors.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.bdd;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import com.puppycrawl.tools.checkstyle.DefaultConfiguration;
+
+public final class ParsedInput {
+
+    /** The check name. */
+    private final String checkName;
+
+    /** Map of properties. */
+    private final Map<String, String> properties;
+
+    /** Map of default properties. */
+    private final Map<String, String> defaultProperties;
+
+    /** Map of non default properties. */
+    private final Map<String, String> nonDefaultProperties;
+
+    /** List of violations. */
+    private final List<String> violations;
+
+    public ParsedInput(String checkName,
+                       Map<String, String> properties,
+                       Map<String, String> defaultProperties,
+                       Map<String, String> nonDefaultProperties,
+                       List<String> violations) {
+        this.checkName = checkName;
+        if (properties == null) {
```

Comment:
lets use builder design pattern for such cases
it is useful constructor have many parameters and 
you can read more here https://dzone.com/articles/design-patterns-the-builder-pattern

Reply:
done

Reply:
@timurt 
PMD complains about constructor. A guy [here](https://stackoverflow.com/questions/6174419/joshua-blochs-builder-pattern-and-pmd-warnings) say that "PMD is just being dumb". Please suggest.

Reply:
This is fixed in PMD 7.0: https://github.com/pmd/pmd/issues/755#issuecomment-745261250
For PMD 6.xx this can be suppressed in the pmd config.

Reply:
done